crates/turborepo-lib/src/process/mod.rs

```diff
@@ -0,0 +1,286 @@
+//! `process`
+//!
+//! This module contains the code that is responsible for running the commands
+//! that are queued by run. It consists of a set of child processes that are
+//! spawned and managed by the manager. The manager is responsible for
+//! running these processes to completion, forwarding signals, and closing
+//! them when the manager is closed.
+//!
+//! As of now, the manager will execute futures in a random order, and
+//! must be either `wait`ed on or `stop`ped to drive state.
+
+mod child;
+
+use std::{
+    sync::{atomic::AtomicBool, Arc, Mutex},
+    time::Duration,
+};
+
+pub use child::Command;
+use futures::Future;
+use tokio::task::JoinSet;
+use tracing::{debug, trace};
+
+use self::child::{Child, ChildExit};
+
+/// A process manager that is responsible for spawning and managing child
+/// processes. When the manager is Open, new child processes can be spawned
+/// using `spawn`. When the manager is Closed, all currently-running children
+/// will be closed, and no new children can be spawned.
+#[derive(Debug, Clone)]
+pub struct ProcessManager {
+    is_closing: Arc<AtomicBool>,
+    children: Arc<Mutex<Vec<child::Child>>>,
```

Comment:
I think to avoid to sidestep the complexity of splitting these two pieces of state into their own individual sync mechanisms, we should just bundle them together behind a shared lock. This is what we do in Go and it protects us from using one of these pieces of information without being synced with the other.
```suggestion
    inner: Arc<Mutex<ManagerInner>>,
```

```
struct ManagerInner {
    is_closing: bool,
    children: Vec<child::Child>,
}
```